### PR TITLE
tests: dm master module add unit test

### PR DESCRIPTION
**Pull request by amyangfei**
### What problem does this PR solve? 
add unit test to dm/master module

### What is changed and how it works?
1. add dm-master config test
2. add dm-master http API test
3. add dm-master rpc API test

* some code change
   - fix worker resp op assignment in OperateWorkerRelayTask
   - always update task-workers mapping in `RefreshWorkerTasks`, will refine this later

### Check List 

Tests 

 - Unit test

### CODE REVIEW

**Comment by amyangfei** [comment]
PTAL @GregoryIan @csuzhangxc

File: dm/master/config_test.go [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+}
+
+func (t *testMaster) TestUpdateConfig(c *check.C) {
+	var (
+		f          *os.File
+		err        error
+		content    []byte
+		newContent []byte
+	)
+	f, err = ioutil.TempFile(".", "test_config.toml")
```
Comment (GregoryIan) [inline dm/master/config_test.go]: `c.Mkdir` is better

File: dm/master/status_test.go [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+}
+
+const retryTime = 100
+
+func (t *testHTTPServer) waitUntilServerOnline() error {
+	statusURL := fmt.Sprintf("http://127.0.0.1%s/status", t.cfg.MasterAddr)
+	for i := 0; i < retryTime; i++ {
+		resp, err := http.Get(statusURL)
+		if err == nil {
+			ioutil.ReadAll(resp.Body)
```
Comment (GregoryIan) [inline dm/master/status_test.go]: check `status code`?

File: dm/master/status_test.go [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+		time.Sleep(time.Millisecond * 10)
+	}
+	return errors.Errorf("failed to connect http status for %d retries in every 10ms", retryTime)
+}
+
+func (t *testHTTPServer) TestStatus(c *check.C) {
+	t.startServer(c)
+	defer t.stopServer(c)
+
+	statusURL := fmt.Sprintf("http://127.0.0.1%s/status", t.cfg.MasterAddr)
```
Comment (GregoryIan) [inline dm/master/status_test.go]: we have checked it in `waitUntilServerOnline`,  lol 🎁

File: dm/master/server_test.go [hunk 4]
```diff
@@ -14,16 +14,1449 @@
...
+func defaultMasterServer(c *check.C) *Server {
+	cfg := NewConfig()
+	err := cfg.Parse([]string{"-config=./dm-master.toml"})
+	c.Assert(err, check.IsNil)
+	server := NewServer(cfg)
+
+	return server
+}
+
+func genSubTaskConfig(c *check.C, server *Server, ctrl *gomock.Controller) map[string]*config.SubTaskConfig {
```
Comment (GregoryIan) [inline dm/master/server_test.go]: should we add a prefix `test` for function of ut?

**Comment by GregoryIan** [comment]
Rest LGTM

File: dm/master/config_test.go [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+	c.Assert(newContent, check.DeepEquals, content)
+}
+
+func (t *testMaster) TestInvalidConfig(c *check.C) {
+	var (
+		f   *os.File
+		err error
+		cfg = NewConfig()
+	)
+	f, err = ioutil.TempFile(".", "test_invalid_config.toml")
```
Comment (WangXiangUSTC) [inline dm/master/config_test.go]: how about use `:=` and don't need var f and err

Comment (amyangfei) [reply dm/master/config_test.go]: use `c.MkDir` and create file in temp dir instead

**Comment by GregoryIan** [comment]
LGTM

File: dm/master/server_test.go [hunk 6]
```diff
@@ -14,16 +14,1470 @@
...
+    worker-count: 16
+    batch: 100
+    max-retry: 100
+`
+
+var (
+	errGRPCFailed      = "test grpc request failed"
+	errExecDDLFailed   = "dm-worker exec ddl failed"
+	msgNoSubTask       = "no sub task started"
+	errCheckSyncConfig = "check sync config with error"
```
[the remark this answers is not in the document]
Comment (amyangfei) [reply dm/master/server_test.go]: These  error messages are mostly used in mock rpc response message, and not really used in dm-master server. I think it is fine to keep them here.

File: dm/master/config_test.go [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+			c.Assert(err, check.ErrorMatches, tc.errorReg)
+		} else {
+			c.Assert(cfg.MasterAddr, check.Equals, masterAddr)
+			c.Assert(cfg.DeployMap, check.DeepEquals, deployMap)
+			c.Assert(cfg.String(), check.Matches, fmt.Sprintf("{.*master-addr\":\"%s\".*}", masterAddr))
+		}
+	}
+}
+
+func (t *testMaster) TestUpdateConfig(c *check.C) {
```
[the remark this answers is not in the document]
Comment (amyangfei) [reply dm/master/config_test.go]: addressed

File: dm/master/server_test.go [hunk 8]
```diff
@@ -14,16 +14,1470 @@
...
+	server.lockKeeper = lk
+
+	// test query with task name
+	resp, err = server.ShowDDLLocks(context.Background(), &pb.ShowDDLLocksRequest{
+		Task:    "testA",
+		Workers: workers,
+	})
+	c.Assert(err, check.IsNil)
+	c.Assert(resp.Result, check.IsTrue)
+	c.Assert(resp.Locks, check.HasLen, 1)
```
[the remark this answers is not in the document]
Comment (amyangfei) [reply dm/master/server_test.go]: check the `resp.Locks[0].ID` value. as we focus the rpc test in this test case, the more specific DDL lock value is not concerned

**Comment by WangXiangUSTC** [comment]
LGTM

**Pull request merged**